### PR TITLE
Cut based photon id run iii winter22

### PR DESCRIPTION
**Pull request by shdutta16**
#### PR description:

This PR implements the CutBased PhotonID RunIII Winter22 in miniAOD and nanoAOD. The main changes are: 1) H/E variable is now cone-based instead of tower-based (as was in Run2 Fall17 ID); 2) PF Photon Isolation and Neutral Hadron isolation variables are now replaced by ECal Cluster Isolation and HCal Cluster Isolation; 3) H/E variable is now pile-up corrected with quadratic polynomial in rho; 4) Pile-up correction with quadratic polynomial in rho is done for isolation variables as well. 
Additional documentation can be found in the following:
1. https://indico.cern.ch/event/1204275/#5-update-on-run3-photon-cut-ba
2. https://indico.cern.ch/event/1204277/#5-update-on-run3-photon-cut-ba  

#### PR validation:

The PR has been tested using the following workflows by running 'runTheMatrix.py'
1) 12434.0    
2) 11834.0    
3) 10024.0    
4) 136.874

### CODE REVIEW

**Comment by swagata87** [comment]
type egamma

**Comment by swertz** [comment]
enable nano

**Comment by swagata87** [comment]
@cmsbuild please test

File: CommonTools/Egamma/src/EffectiveAreas.cc [hunk 1]
```diff
@@ -40,6 +44,48 @@ EffectiveAreas::EffectiveAreas(const std::string& filename) : filename_(filename
   checkConsistency();
 }
 
+EffectiveAreas::EffectiveAreas(const std::string& filename, const bool quadraticEAflag)
```
Comment (swertz) [inline CommonTools/Egamma/src/EffectiveAreas.cc:47]: Shouldn't this constructor also be able to handle the case of `quadraticEAflag=false`, e.g. by calling the other constructor?

File: PhysicsTools/NanoAOD/plugins/HoverEValueMapProducer.cc [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    doHoverEPho(iEvent);
+  }
+}
+
+template <typename T>
+void HoverEValueMapProducer<T>::doHoverEPho(edm::Event& iEvent) const {}
+
+template <>
+void HoverEValueMapProducer<pat::Photon>::doHoverEPho(edm::Event& iEvent) const {
+  edm::Handle<edm::View<pat::Photon>> src;
```
Comment (swertz) [inline PhysicsTools/NanoAOD/plugins/HoverEValueMapProducer.cc:108]: Please use the new style:
```
auto src = iEvent.getHandle(src_);
const double rho = iEvent.get(rho_);
```

File: PhysicsTools/NanoAOD/plugins/HoverEValueMapProducer.cc [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    desc.add<edm::InputTag>("rho")->setComment(
+        "rho to be used for effective-area based H/E pileup subtraction for photons");
+  }
+
+  std::string modname;
+  if (typeid(T) == typeid(pat::Photon))
+    modname += "Pho";
+  modname += "HoverEValueMapProducer";
+  descriptions.add(modname, desc);
```
Comment (swertz) [inline PhysicsTools/NanoAOD/plugins/HoverEValueMapProducer.cc:156]: I think you can just do `descriptions.addWithDefaultLabel(desc);` and it will use the name of the template typedef.

File: PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc [hunk 4]
```diff
@@ -255,8 +288,21 @@ void IsoValueMapProducer<pat::Photon>::doPFIsoPho(edm::Event& iEvent) const {
   fillerAll.insert(src, PFIsoAll.begin(), PFIsoAll.end());
   fillerAll.fill();
 
+  std::unique_ptr<edm::ValueMap<float>> PFIsoChgQuadraticV(new edm::ValueMap<float>());
```
Comment (swertz) [inline PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc:291]: Please use `make_unique` as in https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc#L249

File: PhysicsTools/NanoAOD/python/V10/photons_cff.py [hunk 5]
```diff
@@ -13,6 +13,7 @@
         'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Fall17_94X_V2_cff',
```
Comment (swertz) [inline PhysicsTools/NanoAOD/python/V10/photons_cff.py:13]: Please revert the changes to the V10 config. The V10 is frozen and should not be modified.

File: PhysicsTools/NanoAOD/python/photons_cff.py [hunk 6]
```diff
@@ -122,8 +151,14 @@ def make_bitmapVID_docstring(id_modules_working_points_pset):
         mvaID = cms.InputTag("photonMVAValueMapProducer:PhotonMVAEstimatorRunIIFall17v2Values"),
         PFIsoChg = cms.InputTag("isoForPho:PFIsoChg"),
         PFIsoAll = cms.InputTag("isoForPho:PFIsoAll"),
+        PFIsoChgQuadratic = cms.InputTag("isoForPho:PFIsoChgQuadratic"),
+        PFIsoAllQuadratic = cms.InputTag("isoForPho:PFIsoAllQuadratic"),
+        HoverEQuadratic = cms.InputTag("hOverEForPho:HoEForPhoEACorr"),
```
Comment (swertz) [inline PhysicsTools/NanoAOD/python/photons_cff.py:156]: These 3 user floats are not used anywhere, are they? Don't you want to add them to the photon table?

**Review by swertz — CHANGES_REQUESTED** [review]
Thanks @shdutta16 , I have a few comments on the proposed changes.

File: CommonTools/Egamma/src/EffectiveAreas.cc (hunk 1 above)
Comment (shdutta16) [reply CommonTools/Egamma/src/EffectiveAreas.cc:47]: Yes, indeed. But, still I kept the previous code unchanged. I can delete the other constructor if you suggest so.

File: PhysicsTools/NanoAOD/plugins/HoverEValueMapProducer.cc (hunk 2 above)
Comment (shdutta16) [reply PhysicsTools/NanoAOD/plugins/HoverEValueMapProducer.cc:108]: Ok, thanks. I will make this change and update the branch.

File: PhysicsTools/NanoAOD/plugins/HoverEValueMapProducer.cc (hunk 3 above)
Comment (shdutta16) [reply PhysicsTools/NanoAOD/plugins/HoverEValueMapProducer.cc:156]: Ok, thanks for the suggestion.

File: PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc (hunk 4 above)
Comment (shdutta16) [reply PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc:291]: Ok, will do.

File: PhysicsTools/NanoAOD/python/V10/photons_cff.py (hunk 5 above)
Comment (shdutta16) [reply PhysicsTools/NanoAOD/python/V10/photons_cff.py:13]: The WF 2500.6 was calling the V10 nanoAOD and hence, crashed since it was calling the /V10/photons_cff.py file. I changed the V10 because of that. Should I revert back the changes then?

File: PhysicsTools/NanoAOD/python/photons_cff.py (hunk 6 above)
Comment (shdutta16) [reply PhysicsTools/NanoAOD/python/photons_cff.py:156]: Yes, I will add them there.

File: CommonTools/Egamma/src/EffectiveAreas.cc (hunk 1 above)
Comment (swertz) [reply CommonTools/Egamma/src/EffectiveAreas.cc:47]: It seems you still want to support the case where `quadraticEAflag` is `false`, but right now calling `EffectiveAreas(..., false)` does not produce the right behaviour.
What you could to is put the code from the first constructor into a `LoadLegacyEAFile` method and the code from the second one into a `LoadQuadraticEAFile` method. Then, the first constructor always calls `LoadLegacyEAFile`, while the second one calls the right method depending on the value of `quadraticEAflag`.

Comment (shdutta16) [reply CommonTools/Egamma/src/EffectiveAreas.cc:47]: I was unable to add my response to your latest comment on this, so replying here. I have removed the old constructor and made `quadraticEAflag=false` the default value in the new constructor. So, any portion of the existing code calling this constructor will now produce the correct behaviour.

File: PhysicsTools/NanoAOD/python/V10/photons_cff.py (hunk 5 above)
Comment (shdutta16) [reply PhysicsTools/NanoAOD/python/V10/photons_cff.py]: I have revert back the changes in the V10 nanoAOD and made changes to `IsoValueMapProducer` so, that quadratic EA correction is optional.

**Comment by mandrenguyen** [comment]
please test

**Comment by swagata87** [comment]
Hi, the test seems stuck for 2 days. Is there a way to recover from it?

**Comment by swertz** [comment]
please abort

File: PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc [hunk 7]
```diff
@@ -292,9 +370,27 @@ void IsoValueMapProducer<T>::fillDescriptions(edm::ConfigurationDescriptions& de
     desc.add<edm::FileInPath>("EAFile_PFIso_Pho")
         ->setComment(
             "txt file containing effective areas to be used for photon PF-isolation pileup subtraction for photons");
+
     desc.add<edm::InputTag>("rho_PFIso")
         ->setComment("rho to be used for effective-area based PF-isolation pileup subtraction for photons");
+
+    desc.addOptional<edm::InputTag>("mapIsoEcal")->setComment("input ecal PF isolation calculated in VID for photons");
+    desc.addOptional<edm::InputTag>("mapIsoHcal")->setComment("input hcal PF isolation calculated in VID for photons");
```
Comment (swertz) [inline PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc:378]: Why are these two needed at all? They are nowhere used in the code... And actually, neither are `mapIsoChg`, `mapIsoNeu` and `mapIsoPho` above...

Comment (shdutta16) [reply PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc:378]: Ok, I will remove them all.

**Comment by swertz** [comment]
please test

File: PhysicsTools/NanoAOD/python/nanoDQM_cfi.py [hunk 8]
```diff
@@ -593,6 +593,7 @@
                 Count1D('_size', 9, -0.5, 8.5, 'slimmedPhotons after basic selection (pt > 5 )'),
                 Plot1D('charge', 'charge', 1, -0.5, 0.5, 'electric charge'),
                 Plot1D('cleanmask', 'cleanmask', 1, 0.5, 1.5, 'simple cleaning mask with priority to leptons'),
+                Plot1D('cutBased_RunIIIWinter22','cutBased_RunIIIWinter22', 4, -0.5, 3.5,'cut-based ID bitmap, RunIIIWinter22, (0:fail, 1:loose, 2:medium, 3:tight)'),
```
Comment (swertz) [inline PhysicsTools/NanoAOD/python/nanoDQM_cfi.py:596]: This should be changed to `cutBased_Fall17V2`, following the change in the table

File: PhysicsTools/NanoAOD/python/photons_cff.py [hunk 9]
```diff
@@ -186,7 +221,10 @@ def make_bitmapVID_docstring(id_modules_working_points_pset):
         pfChargedIsoWorstVtx = Var("chargedHadronWorstVtxIso()",float,doc="PF absolute isolation dR=0.3, charged component (Vertex with largest isolation)"),
         pfRelIso03_chg = Var("userFloat('PFIsoChg')/pt",float,doc="PF relative isolation dR=0.3, charged component (with rho*EA PU corrections)"),
         pfRelIso03_all = Var("userFloat('PFIsoAll')/pt",float,doc="PF relative isolation dR=0.3, total (with rho*EA PU corrections)"),
+        pfRelIso03_chg_quadratic = Var("userFloat('PFIsoChgQuadratic')/pt",float,doc="PF relative isolation dR=0.3, charged hadron component (with quadraticEA*rho*rho + linearEA*rho corrections)"),
+        pfRelIso03_all_quadratic = Var("userFloat('PFIsoAllQuadratic')/pt",float,doc="PF relative isolation dR=0.3, total (with quadraticEA*rho*rho + linearEA*rho corrections)"),
         hoe = Var("hadronicOverEm()",float,doc="H over E",precision=8),
+        hoe_PUcorr = Var("userFloat('HoverEQuadratic')",float,doc="PU corrected H/E (cone-based with quadraticEA*rho*rho + linearEA*rho corrections)",precision=8),
```
Comment (swertz) [inline PhysicsTools/NanoAOD/python/photons_cff.py:227]: Could you please also add these 3 new variables to the DQM?

File: PhysicsTools/NanoAOD/python/photons_cff.py [hunk 10]
```diff
@@ -158,21 +184,30 @@ def make_bitmapVID_docstring(id_modules_working_points_pset):
...
+            doc="cut-based ID bitmap, RunIIIWinter22V1, (0:fail, 1:loose, 2:medium, 3:tight)",
+        ),
+        cutBased_Fall17V2 = Var(
+            "userInt('cutbasedID_Fall17V2_loose')+userInt('cutbasedID_Fall17V2_medium')+userInt('cutbasedID_Fall17V2_tight')",
             int,
             doc="cut-based ID bitmap, Fall17V2, (0:fail, 1:loose, 2:medium, 3:tight)",
         ),
-        cutBased_Fall17V1Bitmap = Var(
-            "userInt('cutbasedIDV1_loose')+2*userInt('cutbasedIDV1_medium')+4*userInt('cutbasedIDV1_tight')",
+        cutBased_Fall17V1 = Var(
```
Comment (swertz) [inline PhysicsTools/NanoAOD/python/photons_cff.py:196]: For my curiosity, is the Fall17V1 ID even still supported? Or could it be removed?

**Review by swertz — CHANGES_REQUESTED** [review]
Thanks @shdutta16 , I think we're almost there - just a small fix needed for the nano DQM.

I'd also like to raise that the strategy for the ID in nano is different between electron and photons: in https://github.com/cms-sw/cmssw/pull/40045, the default `cutBased` variable is reverted to Fall17V2 for Run2 samples, and the Run3 ID is not available in Run2 samples. But I'm fine with merging this to have the Run3 ID in for mini, and have a small follow-up PR to make things more uniform in nano.

File: PhysicsTools/NanoAOD/python/photons_cff.py (hunk 10 above)
Comment (shdutta16) [reply PhysicsTools/NanoAOD/python/photons_cff.py:196]: I have removed it

File: PhysicsTools/NanoAOD/python/photons_cff.py (hunk 9 above)
Comment (shdutta16) [reply PhysicsTools/NanoAOD/python/photons_cff.py:227]: Yes, I have added them.

File: PhysicsTools/NanoAOD/python/nanoDQM_cfi.py (hunk 8 above)
Comment (shdutta16) [reply PhysicsTools/NanoAOD/python/nanoDQM_cfi.py:596]: Naming convention have been changed. The way it is done for the mentioned PR

File: PhysicsTools/NanoAOD/python/photons_cff.py (hunk 10 above)
Comment (swagata87) [reply PhysicsTools/NanoAOD/python/photons_cff.py]: Yes just to confirm, Fall17V1 is not supported anymore by EGM for Run2UL and for Run3.
Even for Run2 EOY ReReco, Fall17V1 is disfavoured. We have been encouraging analysers to move to Fall17V2 for years now. So I think it's a good time to remove Fall17V1 from nano.

**Comment by swagata87** [comment]
please test

**Comment by swagata87** [comment]
@swertz  the test that I started yesterday got stuck.. 
maybe you need to give the command for nano tests?

**Comment by swertz** [comment]
please abort

Not that I know of - once nano tests are enabled by anyone, they should run. Tests are probably just stuck, I'll relaunch.

**Comment by swertz** [comment]
please test

**Comment by swertz** [comment]
+1

Changes in nanoAOD photon cutBased ID are expected; new isolation variables are visible in the DQM.

**Comment by mandrenguyen** [comment]
+reconstruction

File: PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc [hunk 11]
```diff
@@ -280,21 +362,39 @@ void IsoValueMapProducer<T>::fillDescriptions(edm::ConfigurationDescriptions& de
...
-    desc.add<edm::FileInPath>("EAFile_PFIso_Neu")
+    desc.addOptional<edm::FileInPath>("EAFile_PFIso_Neu")
         ->setComment(
             "txt file containing effective areas to be used for neutral PF-isolation pileup subtraction for photons");
-    desc.add<edm::FileInPath>("EAFile_PFIso_Pho")
+    desc.addOptional<edm::FileInPath>("EAFile_PFIso_Pho")
         ->setComment(
             "txt file containing effective areas to be used for photon PF-isolation pileup subtraction for photons");
+
     desc.add<edm::InputTag>("rho_PFIso")
```
Comment (perrotta) [inline PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc:381]: I'm curious how do you decide to make a parameter "optional" or not. For example `rho_PFIso` is only used by electrons and photons: why it is not set optional as other parameters not included in all configurations?

There are other such parameters only used for specific particles, which are not set optional: please check that all that is correct

Comment (shdutta16) [reply PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc:381]: The effective Area (EA) correction for `Fall17V2` has different definition (compared to `RunIIIWinter22V1`) and consequently different EA files (.txt) where the values are stored. `QuadraticEAFile_PFIso_` are used only in `RunIIIWinter22V1` while `EAFile_PFIso_` are used only in `Fall17V2`. And that is why parameters corresponding to them are kept optional because only one kind of EA correction is accessed by a particular era of ID. The `mapIso` parameters are neither used in the plugin nor in the photon table. WF like 2500.6 accesses the `V10/nano_cff.py` which has these parameters in its config. Since, we are not touching the V10, we made the `mapIso` parameters optional as well, while not having these parameters in the current `nano_cff.py` config. If I don't do this, WF 2500.6 crashes. `rho_PFIso` is used by both (`Fall17V2` and `RunIIIWinter22V1`) the IDs and hence, it is kept as 'non-optional'.

Comment (perrotta) [reply PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc:381]: @shdutta16 Deciding whether a parameter is Optional or not shouldn't depend on whether it is already present in the existing python configurations. It should rather depend on the structure of the code.

As a possible rule: a parameter should be considered optional if a different behaviour in the code depends on whether the parameter exists or not, or if it is technically possible that such a parameter is not needed in the execution flow (for example if it is only relevant for a given case, e.g. a particle, accessible depending on the value assigned to another configuration parameter). If when a parameter is missing a default value is assigned to it, and the parameter itself with that default value continues to be used in the flow of the code, it is not an optional parameter any more.

If you have different parameters for different particles, I'd say that they should all be considered "optional", unless you specify a default value for them and use it when the parameter is not explicitely listed in the python configuration.

Comment (shdutta16) [reply PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc:381]: If we understand your comment correctly, that is what we are trying to do :

1. The parameters that are optional based on which era of ID is calling `IsoValueMapProducer`, have been kept optional. `QuadraticEAFile_PFIso_` is called by `RunIIIWinter22V1` ID and `EAFile_PFIso_` is called by `Fall17V2` ID. 

2. `rho_PFIso` is essential for Photon and we want the code to throw error if it is not present, and that's why it is not kept optional. 

3. `mapIso` doesn't have any usage in the Run3 scenario, but for nanoAOD V10 they are needed and in a previous comment here, we were advised not to touch V10. So, they are kept optional as well, otherwise V10 workflow will crash which has already been tested.

Comment (perrotta) [reply PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc:381]: Uhm, probably the logic behind is more convoluted than I can imagine from my glances to the code.
Taking into account that those parameters were already as such before your PR, let avoid touching them here.
However, I'm convinced that the whole parameter structure could be simplified, to the ease of future developers and maintainers.
Let merge this as it is now, then.
